asset_dashboard/models.py

```diff
@@ -127,31 +129,80 @@ class Project(models.Model):
+    def save_zones_and_districts(sender, instance, **kwargs):
+        phase_geoms = LocalAsset.get_aggregated_assets_by_phase(instance.phase)
+        instance.phase.project.update_project_zones(instance, phase_geoms)
+        instance.phase.project.update_project_districts(instance, phase_geoms)
+
+    def update_project_zones(self, instance, phase_geoms):
+        zones = Zone.objects.filter(boundary__intersects=phase_geoms)
+
+        # Delete existing relationships so we can have a fresh start.
+        instance.phase.project.zones.clear()
 
         for zone in zones:
             instance.phase.project.zones.add(zone)
             instance.phase.project.save()
 
+    def update_project_districts(self, instance, phase_geoms):
         district_models = [
             ('commissioner_districts', CommissionerDistrict),
             ('senate_districts', SenateDistrict),
             ('house_districts', HouseDistrict)
         ]
 
         for attribute, model in district_models:
-            districts = model.objects.filter(
-                boundary__contains=instance.geom
-            )
+            districts = model.objects.filter(boundary__intersects=phase_geoms)
+
+            project_district = getattr(instance.phase.project, attribute)
+            project_district.clear()
 
             for district in districts:
-                project_district = getattr(instance.phase.project, attribute)
                 project_district.add(district)
                 instance.phase.project.save()
 
 
+class PhaseZoneDistribution(models.Model):
+    phase = models.ForeignKey('Phase', on_delete=models.CASCADE, related_name='phase')
+    zone = models.ForeignKey('Zone', on_delete=models.CASCADE, related_name='zone')
+    zone_distribution_percentage = models.FloatField(null=True, blank=True)
+
+    @receiver([post_save, post_delete], sender='asset_dashboard.LocalAsset')
+    def save_zone_distribution(sender, instance, **kwargs):
+        phase_geoms = LocalAsset.get_aggregated_assets_by_phase(instance.phase)
+
+        total_distribution_area = phase_geoms.area
+
+        distributions_by_zone = LocalAsset.get_distribution_by_zone(phase_geoms)
+
+        zone_percentages = PhaseZoneDistribution.calculate_zone_percentage(
+            distributions_by_zone,
+            total_distribution_area
+        )
+
+        for zone, percentage in zone_percentages.items():
+            zone_distribution, _ = PhaseZoneDistribution.objects.get_or_create(
+                phase=instance.phase,
+                zone=zone
+            )
+
+            zone_distribution.zone_distribution_percentage = percentage
+            zone_distribution.save()
+
+    @classmethod
+    def calculate_zone_percentage(cls, distributions_by_zone, total_distribution_area):
+        percentages_by_zone = {}
+
+        for zone, distribution in distributions_by_zone.items():
+            quotient = distribution / total_distribution_area
+            percent = quotient * 100
```

Comment:
i would leave this out, since later you have to redived by 100 to get the right calculations. if you want to make the percent look good, then you should do that at the edge with a something like this: 

https://www.adamsmith.haus/python/answers/how-to-format-a-number-as-a-percentage-in-python

Reply:
oh cool! so just let it hangout in the database as .75 instead of 75%? makes better sense.